### PR TITLE
ACP-77: Update P-chain state staker tests

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

Factored out of #3487.

1. Rewrites `TestPersistStakers` as `TestState_writeStakers` which uses test tables rather than extensive callbacks.
2. Fixes `TestState_writeStakers` to not break the invariants around subnet validators being a sub-set of the primary network validator set.
3. Expands `TestStateAddRemoveValidator` as `TestState_ApplyValidatorDiffs` which includes additional cases, such as adding an removing validators in the same diff.

## How this works

Generally cleans up existing tests and improves coverage.

## How this was tested

Only tests are modified.

### CODE REVIEW

File: vms/platformvm/state/state_test.go [hunk 1]
```diff
@@ -109,620 +110,384 @@ func TestStateSyncGenesis(t *testing.T) {
...
-			checkValidatorsSet: func(r *require.Assertions, s *state, staker *Staker) {
-				val, err := s.GetCurrentValidator(staker.SubnetID, staker.NodeID)
-				r.NoError(err)
-
-				valsMap := s.validators.GetMap(staker.SubnetID)
-				r.Contains(valsMap, staker.NodeID)
-				valOut := valsMap[staker.NodeID]
-				r.Equal(valOut.NodeID, staker.NodeID)
-				r.Equal(valOut.Weight, val.Weight+staker.Weight)
+			expectedPublicKeyDiff: maybe.Some[*bls.PublicKey](nil),
```
Comment (samliok) [inline vms/platformvm/state/state_test.go:242]: why would adding a validator not update the `publicKeyDiff`?

File: vms/platformvm/state/state_test.go [hunk 2]
```diff
@@ -109,620 +110,384 @@ func TestStateSyncGenesis(t *testing.T) {
...
+		},
+		"add current primary network delegator": {
+			initialStakers:            []*Staker{primaryNetworkCurrentValidatorStaker},
+			initialTxs:                []*txs.Tx{addPrimaryNetworkValidator},
+			staker:                    primaryNetworkCurrentDelegatorStaker,
+			addStakerTx:               addPrimaryNetworkDelegator,
+			expectedCurrentValidator:  primaryNetworkCurrentValidatorStaker,
+			expectedCurrentDelegators: []*Staker{primaryNetworkCurrentDelegatorStaker},
+			expectedValidatorSetOutput: &validators.GetValidatorOutput{
+				NodeID:    primaryNetworkCurrentDelegatorStaker.NodeID,
```
Comment (samliok) [inline vms/platformvm/state/state_test.go:252]: to keep things consistent should this be `primaryNetworkCurrentValidatorStaker.NodeID`? It should be the same value though

File: vms/platformvm/state/state_test.go [hunk 3]
```diff
@@ -109,620 +110,384 @@ func TestStateSyncGenesis(t *testing.T) {
...
+				PublicKey: primaryNetworkCurrentValidatorStaker.PublicKey,
+				Weight:    primaryNetworkCurrentDelegatorStaker.Weight + primaryNetworkCurrentValidatorStaker.Weight,
+			},
+			expectedWeightDiff: &ValidatorWeightDiff{
+				Decrease: false,
+				Amount:   primaryNetworkCurrentDelegatorStaker.Weight,
 			},
```
Comment (samliok) [inline vms/platformvm/state/state_test.go:259]: for these tests why not also sanity check the `expectedPublicKeyDiff`

File: vms/platformvm/state/state_test.go [hunk 4]
```diff
@@ -988,43 +753,49 @@ func TestValidatorWeightDiff(t *testing.T) {
...
+	for i, primaryStaker := range primaryStakers {
+		subnetStakers[i] = Staker{
+			TxID:            ids.GenerateTestID(),
+			NodeID:          primaryStaker.NodeID,
+			PublicKey:       nil, // Key is inherited from the primary network
+			SubnetID:        subnetID,
+			Weight:          uint64(i + 1),
+			StartTime:       primaryStaker.StartTime,
+			EndTime:         primaryStaker.EndTime,
+			PotentialReward: uint64(i + 1),
```
Comment (samliok) [inline vms/platformvm/state/state_test.go:793]: doesn't seem like we use `PotentialReward` anyway, but if start time and end time are the same, then shouldn't the potential reward be 0?

**Review by samliok — APPROVED** [review]

File: vms/platformvm/state/state_test.go (hunk 1 above)
Comment (StephenButtolph) [reply vms/platformvm/state/state_test.go:242]: This is specifying a key diff (providing a `maybe.Some` means there is a key diff present). Adding a validator results in a key diff where the prior value was `nil`.

File: vms/platformvm/state/state_test.go (hunk 3 above)
Comment (StephenButtolph) [reply vms/platformvm/state/state_test.go:259]: These tests are verifying that there is no public key diff.

File: vms/platformvm/state/state_test.go (hunk 4 above)
Comment (StephenButtolph) [reply vms/platformvm/state/state_test.go:794]: > if start time and end time are the same

`endTime` is 24h after `startTime` throughout this test, but you're right these values don't really matter for this test.

**Pull request merged**